demos/JWST/run_eureka_optimization_niriss_SOSS.py

```diff
@@ -0,0 +1,1177 @@
+            continue
+
+        # Otherwise, update the S3 and S4 ECF files using the key as-is.
+        print(f"Updating parameter: {key} : {value}")
+        WriteECF(s3_ECF_path, key, f"{key}\t\t{value}\n")
+        WriteECF(s4_ECF_path, key, f"{key}\t\t{value}\n")
+
+    ## STEP 1.2 - Set save directory for optimized white light curve ##
+    outputdir_WLC_S3 = s3_meta.outputdir
+    outputdir_WLC_S4 = s4_meta.outputdir
+
+    # Save optimized WLC data for reference
+    s3_meta_best_WLC = s3_meta
+    s4_meta_best_WLC = s4_meta
+
+    ## STEP 1.2 - Copy the optimizer_inputs.txt file used to the optimization output folder ##
+    def copy_and_rename_file(src, dest_dir, dest_filename):
+        if not os.path.exists(dest_dir):
+            os.makedirs(dest_dir)
+        shutil.copy(src, os.path.join(dest_dir, dest_filename))
+
+    if __name__ == "__main__":
+        source_file = "optimizer_inputs_niriss_SOSS.txt"
+        destination_directory = outputdir_optimization
+        destination_filename = "optimizer_inputs_niriss_SOSS_copy.txt"
+        
+        copy_and_rename_file(source_file, destination_directory, destination_filename)
+        print(f"'{source_file}' has been copied to '{destination_directory}/{destination_filename}'")
+
+    ## STEP 1.2 - Create output file for optimization results ##
+    output_file = os.path.join(outputdir_optimization, "optimization_results.txt")
+
+    with open(output_file, "w") as f:
+        f.write("## OPTIMIZATION RESULTS ##\n\n")
+        # Write Dictionary Contents
+        f.write("## Optimized ECF & EPF Inputs ##\n")
+        for key, value in best.items():
+            f.write(f"{key}: {value}\n")
+        f.write("\n")
+        # Write Optimization Metrics
+        f.write("## Optimization Metrics ##\n")
+        f.write(f"MAD_WLC = {sum(s4_meta_best_WLC.mad_s4_binned) / len(s4_meta_best_WLC.mad_s4_binned)}\n")
+        f.write(f"MAD_Spec = {s4_meta_best_WLC.mad_s4}\n")
+
+    print(f"Optimization results saved to {output_file}")
+
+    ## STEP 1.2 - Save results for optimized white light curve ##
+    def copy_files_and_subfolders_to_new_directory(src_dirs, dest_dir):
+        if not os.path.exists(dest_dir):
+            os.makedirs(dest_dir)
+        for src_dir, folder_name in src_dirs:
+            if os.path.exists(src_dir):
+                dest_subdir = os.path.join(dest_dir, folder_name)
+                if not os.path.exists(dest_subdir):
+                    os.makedirs(dest_subdir)
+                shutil.copytree(src_dir, dest_subdir, dirs_exist_ok=True)
+
+    directories = [(outputdir_WLC_S3, "S3"), (outputdir_WLC_S4, "S4")]
+    best_directory_white = os.path.join(outputdir_optimization, "White")
+
+    copy_files_and_subfolders_to_new_directory(directories, best_directory_white)
+
+    # Plot Optimization History
+    keys_plot = list(history_fitness_score.keys())
+    values_plot = list(history_fitness_score.values())
+
+    plt.figure(figsize=(12, 6))
```

Comment:
Use a specific figure number (e.g., plt.figure(100, figsize=(12, 6)) and then use that in the file naming scheme ("Fig100_OptimizationHistory.png").